STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob_ustream_interface.c

```diff
@@ -0,0 +1,358 @@
+      size_t size_to_copy = remain_size_inner_buffer < remain_size_local_buffer
+          ? remain_size_inner_buffer
+          : remain_size_local_buffer;
+
+      // copy data from inner buffer to local buffer
+      IGNORE_MEMCPY_TO_NULL
+      memcpy(
+          buffer + (buffer_length - remain_size_local_buffer),
+          blob_http_cb->_internal.packet_ptr->nx_packet_prepend_ptr + ustream_instance->offset_diff,
+          size_to_copy);
+      RESUME_WARNINGS
+
+      // decrease the remaining size in our local buffer
+      remain_size_local_buffer -= size_to_copy;
+
+      // increase the passed size for the caller
+      total_size_copied += size_to_copy;
+
+      // increase offset to keep track of our position in the inner buffer
+      ustream_instance->offset_diff += size_to_copy;
+    }
+
+    // else if we have not already grabbed the last chunk, grab another
+    else if (!blob_http_cb->_internal.last_chunk)
+    {
+      result = az_nx_blob_client_grab_chunk(
+          &blob_http_cb->_internal.http_client, &blob_http_cb->_internal.packet_ptr);
+      if ((result == AZ_OK) || (result == AZ_ULIB_EOF))
+      {
+        // increase current position by size of last chunk
+        ustream_instance->inner_current_position += ustream_instance->offset_diff;
+        // reset current offset
+        ustream_instance->offset_diff = 0;
+      }
+
+      // raise last_chunk flag so we can exit once done processing this last chunk
+      if (result == AZ_ULIB_EOF)
+      {
+        blob_http_cb->_internal.last_chunk = true;
+      }
+    }
+
+    // else this is the last buffer-full to copy, exit
+    else
+    {
+      break;
+    }
+  } while ((remain_size_local_buffer > 0) && ((result == AZ_OK) || (result == AZ_ULIB_EOF)));
+
+  // load total size copied into returned size if successful
+  if ((result == AZ_OK) || (result == AZ_ULIB_EOF))
+  {
+    *size = total_size_copied;
+    result = AZ_OK;
+  }
+
+  // set result to EOF if we have finished processing the last chunk
+  if (blob_http_cb->_internal.last_chunk && (remain_size_inner_buffer <= 0))
+  {
+    result = AZ_ULIB_EOF;
+  }
+
+  return result;
+}
+
+static az_result concrete_get_remaining_size(az_ulib_ustream* ustream_instance, size_t* const size)
+{
+  _az_PRECONDITION(AZ_ULIB_USTREAM_IS_TYPE_OF(ustream_instance, api));
+  _az_PRECONDITION_NOT_NULL(size);
+
+  *size = ustream_instance->length - ustream_instance->inner_current_position;
+
+  return AZ_OK;
+}
+
+static az_result concrete_get_position(az_ulib_ustream* ustream_instance, offset_t* const position)
+{
+  _az_PRECONDITION(AZ_ULIB_USTREAM_IS_TYPE_OF(ustream_instance, api));
+  _az_PRECONDITION_NOT_NULL(position);
+
+  return AZ_ERROR_ULIB_DISABLED;
+}
+
+static az_result concrete_release(az_ulib_ustream* ustream_instance, offset_t position)
+{
+  _az_PRECONDITION(AZ_ULIB_USTREAM_IS_TYPE_OF(ustream_instance, api));
+
+  return AZ_ERROR_ULIB_DISABLED;
+}
+
+static az_result concrete_clone(
```

Comment:
>concrete_clone [](http://example.com/codeflow?start=17&length=14)

This implementation of clone is probably not correct. To clone this blob, you probably need to create a second http connection. On the other hand, I don't think it would be that useful in a ustream_forward, so, I would just make it not available. Because you must implement an `concrete_clone` to complete the ustream interface, the solution is just return `AZ_ERROR_NOT_SUPPORTED` in this function.